### PR TITLE
Implement the data encryption plugin for RocksDB

### PR DESCRIPTION
**Pull request by acelyc111**
Most of the code are migrated from https://github.com/pegasus-kv/rocksdb/commits/v8.5.3-pegasus-encrypt, there is no functional changes, but just refactor to use the plugin architecture [1] of RocksDB.

1. https://github.com/facebook/rocksdb/tree/main/plugin#readme

### CODE REVIEW

File: .github/actions/pre-steps/action.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      echo "SKIP_FORMAT_BUCK_CHECKS=1" >> $GITHUB_ENV
+      echo "GTEST_COLOR=1" >> $GITHUB_ENV
+      echo "CTEST_OUTPUT_ON_FAILURE=1" >> $GITHUB_ENV
+      echo "CTEST_TEST_TIMEOUT=300" >> $GITHUB_ENV
+      echo "ZLIB_DOWNLOAD_BASE=https://rocksdb-deps.s3.us-west-2.amazonaws.com/pkgs/zlib" >> $GITHUB_ENV
+      echo "BZIP2_DOWNLOAD_BASE=https://rocksdb-deps.s3.us-west-2.amazonaws.com/pkgs/bzip2" >> $GITHUB_ENV
+      echo "SNAPPY_DOWNLOAD_BASE=https://rocksdb-deps.s3.us-west-2.amazonaws.com/pkgs/snappy" >> $GITHUB_ENV
+      echo "LZ4_DOWNLOAD_BASE=https://rocksdb-deps.s3.us-west-2.amazonaws.com/pkgs/lz4" >> $GITHUB_ENV
+      echo "ZSTD_DOWNLOAD_BASE=https://rocksdb-deps.s3.us-west-2.amazonaws.com/pkgs/zstd" >> $GITHUB_ENV
+    shell: bash
```
Comment (foreverneverer) [inline .github/actions/pre-steps/action.yml]: new line

**Comment by acelyc111** [comment]
You can compare the changes with PR https://github.com/pegasus-kv/rocksdb/pull/21

**Review by foreverneverer — DISMISSED** [review]

File: README.md [hunk 2]
```diff
@@ -1,2 +1,121 @@
 # encfs
-An encryption plugin for RocksDB.
+
+The `AESEncryptionProvider` of `EncryptionFileSystem` is an encryption plugin for RocksDB. It depends on OpenSSL to show how an external plugin can bring its dependencies into the RocksDB build. It provides a factory function in a header file to show integration with RocksDB header includes. It can also be enabled in text-based options to demonstrate use of the static registration framework.
+
+## Build
+
```
Comment (Smityz) [inline README.md:7]:
```suggestion
Download the RocksDB code from the official repository. Currently, version 8.5.3 is used as the benchmark for testing.

git clone --depth 1 --branch v8.5.3 https://github.com/facebook/rocksdb.git
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,121 @@
...
+Build by cmake and check the functionality:
+```
+#!/usr/bin/env bash
+
+set -ex
+
+# 1. build
+mkdir build && cd build
+cmake -DWITH_LZ4=1 -DCMAKE_BUILD_TYPE=Debug -DWITH_TESTS=1 -DROCKSDB_BUILD_SHARED=0 -DROCKSDB_PLUGINS=encfs ..
+make -j32
```
Comment (Smityz) [inline README.md]:
```suggestion
make -j $(nproc)
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,121 @@
...
+      "id=AES;hex_instance_key=0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF;method=AES192CTR"
+      "id=AES;hex_instance_key=0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF;method=AES256CTR"
+      "id=AES;hex_instance_key=0123456789ABCDEF0123456789ABCDEF;method=SM4CTR")
+for opt in ${opts[*]}; do
+  echo "${opt}"
+
+  # 3.1. Run all tests
+  # Set env ENCRYPTED_ENV and run all tests
+  export ENCRYPTED_ENV=${opt}
+  ctest -j32 -V
```
Comment (Smityz) [inline README.md]:
```suggestion
  ctest -j $(nproc) -V
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,121 @@
...
+  export TEST_FS_URI=${uri}
+  ./env_basic_test --gtest_filter=*CustomEnv*
+  ./env_test --gtest_filter=CreateEnvTest.CreateEncryptedFileSystem
+  unset TEST_FS_URI
+
+  # 3.3. Run benchmarks
+  ./db_bench --fs_uri="${uri}" --benchmarks="fillseq,readrandom,readseq" --compression_type=lz4 --num=1000000
+
+  # 3.4. Run ldb tools
+  ./tools/ldb --fs_uri="${uri}" --db=/tmp/rocksdbtest-1000/dbbench/ put k v
```
Comment (Smityz) [inline README.md:68]: much better to add a hint of installing ldb

File: README.md [hunk 6]
```diff
@@ -1,2 +1,121 @@
...
+  if (!s.ok()) {
+    fprintf(stderr, "WriteStringToFile() failed with error %s\n", s.ToString().c_str());
+    return 1;
+  }
+
+  return 0;
+}
+EOF
+$ g++ -std=c++17 -o test test.cpp -lrocksdb -lpthread -llz4 -lcrypto -u encfs_reg
+$ ./tmp
```
Comment (Smityz) [inline README.md]:
```suggestion
$ ./test
```

File: README.md (hunk 5 above)
Comment (acelyc111) [reply README.md:68]: Both the ldb and db_bench tools are built though the commands above, similar to the test binaries, so I think it's not necessary to mention that specially.

Comment (Smityz) [reply README.md:68]: During testing, I noticed that the file was missing from the directory. It is possible that this is due to a problem with my build, therefore, I will double-check and investigate further.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,123 @@
 # encfs
-An encryption plugin for RocksDB.
+
+The `AESEncryptionProvider` of `EncryptionFileSystem` is an encryption plugin for RocksDB. It depends on OpenSSL to show how an external plugin can bring its dependencies into the RocksDB build. It provides a factory function in a header file to show integration with RocksDB header includes. It can also be enabled in text-based options to demonstrate use of the static registration framework.
+
+## Build
+Download the RocksDB code from the official repository. Currently, version 8.5.3 is used as the benchmark for testing.
+
+git clone --depth 1 --branch v8.5.3 https://github.com/facebook/rocksdb.git
```
Comment (Smityz) [inline README.md:9]: need to add a code block, there are some bugs when I added last suggestion

File: README.md (hunk 5 above)
Comment (acelyc111) [reply README.md:68]: These options in CMakeLists.txt are ON by default:
```
option(WITH_BENCHMARK_TOOLS "build with benchmarks" ON)
option(WITH_CORE_TOOLS "build with ldb and sst_dump" ON)
option(WITH_TOOLS "build with tools" ON)
```

**Review by Smityz — APPROVED** [review]

**Review by empiredan — APPROVED** [review]

**Pull request merged**